### PR TITLE
Refactor milestoning code

### LINKED ISSUE #1148: Bug: Fix rewriting of queries for milestoning
### Similar issues

- [X] I have searched and found no existing similar issues

### How are you using Studio?

Legend Query

### Current and expected behavior

We shouldn't re-write the user given queries when we go back from text mode to form mode.

### Steps to reproduce

Copy paste the below query in query builder.
```
{processingDate: Date[1],businessDate: Date[1]|my::Class1.all($processingDate, $businessDate)->project(
  [
    x|$x.biTemporal(%2021-11-12, %2021-11-12).processingtemporal.firmID
  ],
  ['Bi Temporal/Processingtemporal/Firmid']
)}
```

We re-write the query to
```diff
{processingDate: Date[1],businessDate: Date[1]|my::Class1.all($processingDate, $businessDate)->project(
  [
-   x|$x.biTemporal(%2021-11-12, %2021-11-12).processingtemporal.firmID
+   x|$x.biTemporal(%2021-11-12, %2021-11-12).processingtemporal($processingDate).firmID
  ],
  ['Bi Temporal/Processingtemporal/Firmid']
)}
```
But that's not the correct behaviour. Going back from form mode to text mode shouldn't re-write the query.

### Model data

```text
###Relational
Database my::db
(
  Table PersonTable
  (
    NAME CHAR(200) PRIMARY KEY,
    FIRMID INTEGER
  )
  Table FirmTable
  (
    FIRMNAME CHAR(200) PRIMARY KEY,
    ID INTEGER
  )

  Join Firm_Person(PersonTable.FIRMID = FirmTable.ID)
)


###Service
Service my::service1
{
  pattern: '/f8b86e4e-6b84-4de7-aefd-94bccf23f90b';
  documentation: '';
  autoActivateUpdates: true;
  execution: Single
  {
    query: {businessDate: StrictDate[1],processingDate: StrictDate[1]|my::Person1.all($processingDate, $businessDate)};
    mapping: my::map;
    runtime:
    #{
      mappings:
      [
        my::map
      ];
    }#;
  }
  test: Single
  {
    data: '';
    asserts:
    [
    ];
  }
}

Service my::service_class1
{
  pattern: '/f2e2815f-1fe0-47ee-b61c-e5f1d80d72d1';
  documentation: '';
  autoActivateUpdates: true;
  execution: Single
  {
    query: |my::Class1.all(%latest, %latest);
    mapping: my::map;
    runtime:
    #{
      mappings:
      [
        my::map
      ];
    }#;
  }
  test: Single
  {
    data: '';
    asserts:
    [
    ];
  }
}

Service my::service2
{
  pattern: '/13ba6799-9f9f-430b-9c0d-6855b0ac4c3e';
  documentation: '';
  autoActivateUpdates: true;
  execution: Single
  {
    query: processingDate: StrictDate[1]|my::Person2.all($processingDate);
    mapping: my::map;
    runtime:
    #{
      mappings:
      [
        my::map
      ];
    }#;
  }
  test: Single
  {
    data: '';
    asserts:
    [
    ];
  }
}

Service my::service
{
  pattern: '/e70d2c08-4169-44f1-89d4-0fa9971c3995';
  documentation: '';
  autoActivateUpdates: true;
  execution: Single
  {
    query: businessDate: StrictDate[1]|my::Person.all($businessDate);
    mapping: my::map;
    runtime:
    #{
      mappings:
      [
        my::map
      ];
    }#;
  }
  test: Single
  {
    data: '';
    asserts:
    [
    ];
  }
}


###Pure
Class <<temporal.businesstemporal>> my::Person
{
  name: String[1];
  firmID: Integer[1];
}

Class <<meta::pure::profiles::temporal.bitemporal>> my::Class1
{
  businessTemporal: my::Person[1];
  non_milestoning: my::class[1];
  biTemporal: my::Person1[1];
  processingTemporal: my::Person2[1];
}

Class <<temporal.bitemporal>> my::Person1
{
  name: String[1];
  firmID: Integer[1];
  processingtemporal: my::Person2[1];
}

Class <<temporal.processingtemporal>> my::Person2
{
  name: String[1];
  firmID: Integer[1];
  bitemporal: my::Person1[1];
}

Class my::Firm
{
  legalName: String[1];
  id: Integer[1];
}

Class my::class
{
  name: String[1];
}


###Mapping
Mapping my::map
(
  *my::Person: Pure
  {
    name: 'name',
    firmID: 101
  }
  *my::Person1: Relational
  {
    ~primaryKey
    (
      [my::db]PersonTable.NAME
    )
    ~mainTable [my::db]PersonTable
    processingtemporal[my_Person2]: [my::db]PersonTable.FIRMID,
    firmID: [my::db]PersonTable.FIRMID
  }
  *my::Person2: Relational
  {
    ~primaryKey
    (
      [my::db]PersonTable.NAME
    )
    ~mainTable [my::db]PersonTable
    bitemporal[my_Person1]: [my::db]PersonTable.FIRMID,
    firmID: [my::db]PersonTable.FIRMID
  }
  *my::Class1: Relational
  {
    ~primaryKey
    (
      [my::db]PersonTable.NAME
    )
    ~mainTable [my::db]PersonTable
    businessTemporal[my_Person]: [my::db]PersonTable.FIRMID,
    non_milestoning[my_class]: [my::db]PersonTable.NAME,
    biTemporal[my_Person1]: [my::db]PersonTable.NAME,
    processingTemporal[my_Person2]: [my::db]PersonTable.NAME
  }
  *my::class: Pure
  {
    name: 'name'
  }
)
```

### Environment

```text
Windows
```

### Possible solution and workaround

_No response_

### Contribution

- [x] I would like to work on the fix for this issue

### PR DESCRIPTION
**Pull request by gayathrir11**
## Summary

Closes #1148 

- [x] Fixes rewriting of milestoning queries (#1148)
- [x] Organises the milestoning code in query builder with class implementation for each milestoning stereotype
- [x] Re-write the milestoning logic so that if in a property chain date propagation is supported `INTERNAL__PropagatedValue` will point to the milestoning parameter value of preceding property instead of global milestoning dates

https://user-images.githubusercontent.com/87973126/198974744-5592bd40-cfe0-48f0-9f05-26352019accf.mp4

## How did you test this change?

- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

File: packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx [hunk 1]
```diff
@@ -205,6 +226,75 @@ export const QueryBuilderPropertyExpressionEditor = observer(
     const { propertyExpressionState } = props;
     const handleClose = (): void =>
       propertyExpressionState.setIsEditingDerivedProperty(false);
+    const queryBuilderState = propertyExpressionState.queryBuilderState;
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx:229]: What were we doing before this change? were we resetting without the user confirmation?

File: packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx [hunk 2]
```diff
@@ -205,6 +226,75 @@ export const QueryBuilderPropertyExpressionEditor = observer(
...
+    // Resets the next property expression in the property chain for milestoned properties when the user changes to
+    // change the milestoned dates of current proeprty expression and it propagates those dates to next property expression.
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx:231]: type `proeprty`

File: packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx [hunk 3]
```diff
@@ -205,6 +226,75 @@ export const QueryBuilderPropertyExpressionEditor = observer(
...
+          milestoningStereotype &&
+          parameter instanceof INTERNAL__PropagatedValue &&
+          !matchMilestoningParameterValue(
+            milestoningStereotype,
+            index,
+            parameter.getValue(),
+            queryBuilderState.milestoningState,
+          )
+        ) {
+          const newParameterValue = queryBuilderState.milestoningState
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx:253]: isnt the newParameterValue the value of the current parameter value for the current property?

File: .changeset/weak-lies-relax.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@finos/legend-query-builder': minor
+---
+
+Refactor milestoning code
```
Comment (MauricioUyaguari) [inline .changeset/weak-lies-relax.md:5]: Dont think we need a changeset for the refactor bit but we probably do need one for the confirmation message.

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]
The refactor looks fine. Could you walk me through the confirmation changes? 
Also could we do a similar refactor to the legend-graph milestone builders?

File: packages/legend-query-builder/src/stores/milestoning/QueryBuilderMilestoningState.ts [hunk 5]
```diff
@@ -165,6 +151,23 @@ export class QueryBuilderMilestoningState implements Hashable {
     return milestoningParameter;
   }
 
+  getMilestoningBuilderHelper(
```
Comment (akphi) [inline packages/legend-query-builder/src/stores/milestoning/QueryBuilderMilestoningState.ts:154]: see my comment above, I believe we can chuck this inside of `QueryBuilderMilestoningImplementation`

File: packages/legend-query-builder/src/stores/milestoning/QueryBuilderMilestoningHelper.ts [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+/**
+ * Checks if INTERNAL_PropagatedValue we passed for milestoned properties can be removed or not.
+ * We try to make analysis on whether date propagations is supported for this particular property or not:
+ * 1. If it is supported we check if the value it has is same as global milestoning dates
+ *    (i) Yes -> We replace it with actual ValueSpecification (We assune that here it is not pointing to the preceeding property so date
+ *        propagation cannot be done).
+ *    (ii) No -> We just remove it.
+ * 2. If not supported we replace it with the actual ValueSpecification.
+ */
+export const checkIfInternalPropagatedValueCanBeRemoved = (
```
Comment (akphi) [inline packages/legend-query-builder/src/stores/milestoning/QueryBuilderMilestoningHelper.ts:215]: --> `isPropagatedValueRemovable`

When I think about propagated value (i.e. the bubble), I always think about a generic structure, so I really dislike this kind of method where we have differentiating behavior that pertains to milestoning only like this. What do you think about having a simple boolean flag inside of the propagated value to indicate if we should squash it or pop it. This way, the differentiating logic only lies at where we create the bubble or edit the bubble. When we build value specification, there's no special thing we need to do (I think)

**Review by akphi — CHANGES_REQUESTED** [review]
Overall, this PR really makes the code cleaner by moving most milestoning logic out. Really sorry about my scattered reviews, I'm constantly interupted so it took me a while to finish it; but I have 3 main points:
1. The cleaning up and refactor to isolate each milestoning implementation is great, but we can even push for more, your pattern of `....getMilestoningBuilderHelper(stereotype)...` can almost be applied everywhere and thus, there's really no need for us to have a switch case on MILESTONING_STEREOTYPE anywhere
2. I would love to have `PropagatedValue` to be a generic concept, it's proving to be more useful and generic than it seems to me: right now the way it is written, I could see that it's really meant for milestoning. The way to make it generic to me is not too complex (unless I miss something), we can have a boolean flag to indicate if the propagated value should be popped or squashed, this is really the only differentiating behavior AFAIK, and we can set this flag while on the code path for milestoning handling logic, there should not be any differentiating logic when building the value specification
3. I think what we discussed about regarding the change when user modifies the date makes a lot of sense and I'm glad we go to this now. This flow now makes a lot of sense to me! However, like I commented, we need to slightly shift our thinking a bit here. The main thing is to make people understand that date-propagation is natural, if they want the `easy way out` which is to use global value, that's something they can opt-in

Besides, could you check with this, do we fix any of the bugs mentioned in https://github.com/finos/legend-studio/issues/1149

And last but not least, this is a fairly complex PR - took me a while to get into the right context, let's capture the essence of your change in the PR description. Thanks!!!

File: packages/legend-query-builder/src/stores/milestoning/QueryBuilderMilestoningHelper.ts [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+import {
+  guaranteeNonNullable,
+  guaranteeType,
+  UnsupportedOperationError,
+} from '@finos/legend-shared';
+import type { QueryBuilderMilestoningState } from './QueryBuilderMilestoningState.js';
+import type { QueryBuilderDerivedPropertyExpressionState } from '../QueryBuilderPropertyEditorState.js';
+import type { QueryBuilderState } from '../QueryBuilderState.js';
+
+export const getDerivedPropertyMilestoningSteoreotype = (
```
[the remark this answers is not in the document]
Comment (gayathrir11) [reply packages/legend-query-builder/src/stores/milestoning/QueryBuilderMilestoningHelper.ts:41]: I don't think so. The logic in `QueryBuilderMilestoningImplementation` is useful once you know what milestoning stereotype it is. But here we are trying to find the stereotype

File: packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx [hunk 8]
```diff
@@ -108,6 +119,78 @@ const DerivedPropertyParameterValueEditor = observer(
...
+      nextExpression: ValueSpecification | undefined,
+    ): void => {
+      if (
+        nextExpression instanceof AbstractPropertyExpression &&
+        isDefaultDatePropagationSupported(nextExpression, queryBuilderState) &&
+        nextExpression.func.value.genericType.value.rawType instanceof Class
+      ) {
+        queryBuilderState.applicationStore.setActionAlertInfo({
+          message:
+            'You have just changed a milestoning date in the property expression chain, this date will be propagated down the rest of the chain. Do you want to proceed? Otherwise, you can choose to propagate the global milestoning dates instead.',
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx:175]: I don't like the mention of "global". Seems more of a code term.
Is it right to say "current" or the Main class date or is that not accurate

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]
Minor changes requested. Just worried about using the word "global" not sure if we can think of something else for the modal.

File: packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx (hunk 8 above)
Comment (gayathrir11) [reply packages/legend-query-builder/src/components/QueryBuilderPropertyExpressionEditor.tsx:175]: May be default milestoning dates?

**Review by MauricioUyaguari — APPROVED** [review]

**Review by MauricioUyaguari — APPROVED** [review]

**Pull request merged**